hal/inc/ble_hal.h

```diff
@@ -15,190 +15,747 @@
+        hal_ble_gattc_on_char_disc_evt_t char_disc;
+        hal_ble_gatt_on_data_evt_t data_rec;
+    } params;
+} hal_ble_evts_t;
+
+typedef void (*on_ble_evt_cb_t)(const hal_ble_evts_t* event, void* context);
+typedef void (*on_ble_scan_result_cb_t)(const hal_ble_gap_on_scan_result_evt_t* result, void* context);
+typedef void (*on_ble_disc_service_cb_t)(const hal_ble_gattc_on_svc_disc_evt_t* event, void* context);
+typedef void (*on_ble_disc_char_cb_t)(const hal_ble_gattc_on_char_disc_evt_t* event, void* context);
 
 #ifdef __cplusplus
 extern "C" {
 #endif
 
-// BLE error codes
-typedef enum ble_error {
-    BLE_ERROR_UNKNOWN = SYSTEM_ERROR_UNKNOWN,
-    BLE_ERROR_INVALID_PARAM = SYSTEM_ERROR_INVALID_ARGUMENT,
-    BLE_ERROR_INVALID_STATE = SYSTEM_ERROR_INVALID_STATE,
-    BLE_ERROR_NO_MEMORY = SYSTEM_ERROR_NO_MEMORY,
-    BLE_ERROR_BUSY = SYSTEM_ERROR_BUSY
-} ble_error;
-
-// BLE events
-typedef enum ble_event {
-    BLE_EVENT_ADVERT_STARTED = 1,
-    BLE_EVENT_ADVERT_STOPPED = 2,
-    BLE_EVENT_CONNECTED = 3,
-    BLE_EVENT_DISCONNECTED = 4,
-    BLE_EVENT_CONN_PARAM_CHANGED = 5,
-    BLE_EVENT_CHAR_PARAM_CHANGED = 6,
-    BLE_EVENT_DATA_SENT = 7,
-    BLE_EVENT_DATA_RECEIVED = 8
-} ble_event;
-
-// Profile flags
-typedef enum ble_profile_flag {
-    BLE_PROFILE_FLAG_ENABLE_PAIRING = 0x01
-} ble_profile_flag;
-
-// Characteristic flags
-typedef enum ble_char_flag {
-    BLE_CHAR_FLAG_REQUIRE_PAIRING = 0x01
-} ble_char_flag;
-
-// Flags for ble_set_char_value()
-typedef enum ble_set_char_value_flag {
-    BLE_SET_CHAR_VALUE_FLAG_NOTIFY = 0x01
-} ble_set_char_value_flag;
-
-// Characteristic types
-// TODO: Provide an API for custom characteristics and permissions
-typedef enum ble_char_type {
-    BLE_CHAR_TYPE_TX = 1,
-    BLE_CHAR_TYPE_RX = 2,
-    BLE_CHAR_TYPE_VAL = 3
-} ble_char_type;
-
-// UUID
-typedef struct ble_uuid {
-    uint8_t type; // UUID type obtained via ble_add_base_uuid()
-    uint8_t reserved;
-    uint16_t uuid;
-} ble_uuid;
-
-// Characteristic
-typedef struct ble_char {
-    ble_uuid uuid;
-    uint16_t type; // See `ble_char_type` enum
-    uint16_t handle; // TODO: Use typedefs for all handle types defined by the HAL
-    const char* data; // Initial value
-    uint16_t size;
-    uint16_t flags;
-} ble_char;
-
-// Service
-typedef struct ble_service {
-    ble_uuid uuid;
-    ble_char* chars;
-    uint16_t char_count;
-} ble_service;
-
-// Manufacturer-specific data
-typedef struct ble_manuf_data {
-    uint16_t company_id;
-    uint16_t size;
-    const char* data;
-} ble_manuf_data;
-
-// BLE_EVENT_CONNECTED event data
-typedef struct ble_connected_event_data {
-    uint16_t conn_handle;
-} ble_connected_event_data;
+/**
+ * Initialize the BLE stack. This function must be called previous to any other BLE APIs.
+ *
+ * @param[in]   reserved    Reserved for future use.
+ *
+ * @returns     0 on success, system_error_t on error.
+ */
+int ble_stack_init(void* reserved);
```

Comment:
Nitpicking: I'd prefer for us to settle on the common naming scheme between the types and function names. Here we have `hal_ble_xxx` types and `ble_xxx` function names. Not saying that we necessarily need to do this, but just something to consider.